File: Algorithms/Longest Palindromic Substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.length();
        int mx = 1, idx = 0;
        
        bool dp[n][n];
        memset(dp, false, sizeof(dp));
        
        // length 1
        for(int i = 0; i < n; i++)
            dp[i][i] = true;
        
        // length 2    
        for(int i = 0; i < n - 1; i++) {
            if(s[i] == s[i + 1]) {
                dp[i][i + 1] = true;
                idx = i;
                mx = 2;
            }
        }
        
        // k length of palindrome
        for(int k = 3; k <= n; k++)
        {
            // fix the starting
            for(int i = 0; i < n - k + 1; i++)
            {
                // ending index
                int j = i + k - 1;
                if(dp[i + 1][j - 1] && s[i] == s[j]) {
                    dp[i][j] = true;
 
                    if (k > mx) {
                        idx = i;
                        mx = k;
                    }
                }
            }
        }
        
        string ans = "";
        for(int i = idx; i < mx + idx; i++)
            ans += s[i];
        
        return ans;
    }
};

File: January, 2021 Challenge/Longest Palindromic Substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        mx = 1
        idx = 0
        
        dp = [[False for _ in range(n)] for _ in range(n)]
        
        # length 1
        for i in range(n):
            dp[i][i] = True
        
        # length 2
        for i in range(n - 1):
            if s[i] == s[i + 1]:
                dp[i][i + 1] = True
                mx = 2
                idx = i
                
        # length k
        for k in range(3, n + 1):
            # fix the start
            for i in range(n - k + 1):
                j = i + k - 1
                if dp[i + 1][j - 1] and s[i] == s[j]:
                    dp[i][j] = True
                    
                    if k > mx:
                        mx = k
                        idx = i
        
        ans = ""
        for i in range(idx, idx + mx):
            ans += s[i]
    
        return ans
        
